### PR TITLE
To support delete in toss.

### PR DESCRIPTION
**Pull request by liuyu85cn**
#### What type of PR is this?
- [ ] bug
- [X ] feature
- [ ] enhancement

#### Which issue(s)/PR(s) this PR relates to?

support delete out/in edges in TOSS. 

  
#### Checklist：
- [ ] Documentation affected （Please add the label if documentation needs to be modified.)
- [ ] Incompatible （If it is incompatible, please describe it and add corresponding label.）
- [ ] Need to cherry-pick （If need to cherry-pick to some branches, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

#### Release notes：
Please confirm whether to reflect in release notes and how to describe:
>                                                                 `

### CODE REVIEW

File: src/storage/transaction/ChainDeleteEdgesGroupProcessor.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                      : resp.get_result().get_failed_parts().begin()->get_code();
+      handleAsync(spaceId, localPartId, code);
+    });
+    proc->process(request.second);
+  };
+
+  std::for_each(splitedRequest.begin(), splitedRequest.end(), fnSplit);
+}
+
+StatusOr<SplitedRequest> ChainDeleteEdgesGroupProcessor::splitRequest(
```
Comment (panda-sheep) [inline src/storage/transaction/ChainDeleteEdgesGroupProcessor.cpp:46]: Why not keep it consistent with the `shuffleRequest` in `ChainAddEdgesGroupProcessor`?

File: src/clients/storage/InternalStorageClient.cpp [hunk 2]
```diff
@@ -131,5 +131,46 @@ cpp2::ChainAddEdgesRequest InternalStorageClient::makeChainAddReq(const cpp2::Ad
...
+                                             const std::string& txnId,
+                                             TermID termId,
+                                             folly::Promise<nebula::cpp2::ErrorCode>&& p,
+                                             folly::EventBase* evb) {
+  auto spaceId = req.get_space_id();
+  auto partId = req.get_parts().begin()->first;
+  auto optLeader = getLeader(req.get_space_id(), partId);
+  if (!optLeader.ok()) {
+    LOG(WARNING) << folly::sformat("failed to get leader, space {}, part {}", spaceId, partId);
+    p.setValue(::nebula::cpp2::ErrorCode::E_SPACE_NOT_FOUND);
```
Comment (panda-sheep) [inline src/clients/storage/InternalStorageClient.cpp:144]: The error code in this place is not appropriate, including `InternalStorageClient::chainAddEdges`

File: src/storage/transaction/ChainDeleteEdgesLocalProcessor.cpp [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+namespace storage {
+
+void ChainDeleteEdgesLocalProcessor::process(const cpp2::DeleteEdgesRequest& req) {
+  auto rc = checkRequest(req);
+  if (rc != Code::SUCCEEDED) {
+    pushResultCode(rc, localPartId_);
+    finish();
+    return;
+  }
+  env_->txnMan_->addChainTask(this);
```
Comment (panda-sheep) [inline src/storage/transaction/ChainDeleteEdgesLocalProcessor.cpp:27]: 👍，the same logic as addedge

File: src/storage/transaction/ChainDeleteEdgesGroupProcessor.cpp (hunk 1 above)
Comment (liuyu85cn) [reply src/storage/transaction/ChainDeleteEdgesGroupProcessor.cpp:46]: I start to like StartOr ...

File: src/clients/storage/InternalStorageClient.cpp (hunk 2 above)
Comment (liuyu85cn) [reply src/clients/storage/InternalStorageClient.cpp:144]: actually, the status returned may be "space_not_found", or "part_not_found". 
normally, neither would happen here. 
(if this really happen, we can't get here, should stop before)
But we may occur drop space during toss execution, thus I return space_not_found.

**Review by panda-sheep — APPROVED** [review]
Good job!

File: src/storage/transaction/ChainDeleteEdgesGroupProcessor.cpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+using ChainID = std::pair<PartitionID, PartitionID>;
+using SplitedRequest = std::unordered_map<ChainID, cpp2::DeleteEdgesRequest>;
+
+void ChainDeleteEdgesGroupProcessor::process(const cpp2::DeleteEdgesRequest& req) {
+  auto spaceId = req.get_space_id();
+  auto localPartId = req.get_parts().begin()->first;
+  auto stSplitRequest = splitRequest(req);
+  if (!stSplitRequest.ok()) {
+    // TODO(liuyu): change this when error code done
+    pushResultCode(Code::E_SPACE_NOT_FOUND, localPartId);
```
[the remark this answers is not in the document]
Comment (liuyu85cn) [reply src/storage/transaction/ChainDeleteEdgesGroupProcessor.cpp:24]: ok

File: src/storage/transaction/ChainDeleteEdgesLocalProcessor.cpp [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+  doRpc(std::move(pro), std::move(reversedRequest));
+  return std::move(fut);
+}
+
+folly::SemiFuture<Code> ChainDeleteEdgesLocalProcessor::processLocal(Code code) {
+  VLOG(1) << txnId_ << " processRemote(), code = " << apache::thrift::util::enumNameSafe(code);
+
+  bool remoteFailed{false};
+  if (code == Code::SUCCEEDED) {
+    // do nothing
```
[the remark this answers is not in the document]
Comment (liuyu85cn) [reply src/storage/transaction/ChainDeleteEdgesLocalProcessor.cpp:74]: yes.

File: src/storage/transaction/ChainDeleteEdgesLocalProcessor.cpp [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+ * |                   | input keys               | input a batch                        |
+ * |-------------------|--------------------------|------------------------------------|
+ * | double prime (N)  | del edge, prime keys     | bat.remove(prime)               |
+ * | double prime (Y)  | transform to batchHolder | bat.remove(prime) & put(double p.) |
+ */
+void ChainDeleteEdgesLocalProcessor::hookFunc(HookFuncPara& para) {
+  std::string ret;
+
+  if (setDoublePrime_) {
+    if (para.keys) {
```
[the remark this answers is not in the document]
Comment (liuyu85cn) [reply src/storage/transaction/ChainDeleteEdgesLocalProcessor.cpp:252]: yes.

File: src/storage/transaction/ChainDeleteEdgesLocalProcessor.cpp [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+  if (!checkTerm(localPartId_, term_)) {
+    LOG(WARNING) << "E_OUTDATED_TERM";
+    code_ = Code::E_OUTDATED_TERM;
+  }
+
+  if (code == Code::E_RPC_FAILURE) {
+    for (auto& kv : primes_) {
+      auto key =
+          ConsistUtil::doublePrimeTable().append(kv.first.substr(ConsistUtil::primeTable().size()));
+      doublePrimes_.emplace_back(key, kv.second);
```
[the remark this answers is not in the document]
Comment (liuyu85cn) [reply src/storage/transaction/ChainDeleteEdgesLocalProcessor.cpp:93]: fixed.

File: src/storage/transaction/ChainDeleteEdgesLocalProcessor.cpp [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+  if (code == Code::SUCCEEDED) {
+    // do nothing
+  } else if (code == Code::E_RPC_FAILURE) {
+    code_ = Code::SUCCEEDED;
+  } else {
+    code_ = code;
+    remoteFailed = true;
+  }
+
+  if (!checkTerm(localPartId_, term_)) {
```
[the remark this answers is not in the document]
Comment (liuyu85cn) [reply src/storage/transaction/ChainDeleteEdgesLocalProcessor.cpp]: ok

**Review by panda-sheep — APPROVED** [review]
Good job!

**Review by critical27 — APPROVED** [review]

**Pull request merged**